### ISSUE TITLE
Better error output for await in non-async functions

### ISSUE THREAD

**Issue by danez** [babel/babylon#113]
> Issue originally made by @SystemParadox on babel/babel#2387

``` js
function test() {
    await foo();
}
```

Throws: `Unexpected token(2:8)`

This is a bit misleading. Can we change this to say something like "await can only be used in async functions"?

Thanks.

**Comment by motiz88** [comment]
Hi! I'm trying to implement this and learn some Babylon internals at the same time.

**tl;dr: I have a question about a specific implementation idea - see the very end**

### :thought_balloon: Larger discussion: `await` spec

The spec says [here](https://tc39.github.io/ecmascript-asyncawait/#async-function-definitions) under note 1 that (emphasis mine)

> `yield` is _Identifier_ or _YieldExpression_ [ ... ] depending on context. **`await` is always parsed as an _AwaitExpression_ but is a Syntax Error via static semantics.** [ ... ]

Does this mean we can skip trying to disambiguate `await` and just treat it as starting an AwaitExpression everywhere (i.e. immediately an error if not in an async context)? Probably not. I must be missing something. But it would make fixing this issue quite trivial. Moving on...

### :point_right: The issue at hand

I'm assuming the intent here is to optimistically parse `await` as Identifier in non-async contexts, the way it's already being done. But then if the parse turns out to be a dead end, we should "remember" this choice and emit a descriptive error on the `await` - not at the dead end.

I came up with the following logic:

1. When an Identifier with `name === "await"` is parsed, if we didn't enter `parseAwait` because we're in a non-async context, save the identifier's position in `state.potentialIllegalAwaitAt` and keep parsing normally.
2. When parsing a node fails via `unexpected()`, and if `state.potentialIllegalAwaitAt >= node.start`, raise an "await can only be used in async function" error instead of the default "unexpected token".

I'm not 100% sure how to intercept the `unexpected()` occurrences elegantly while still having access to the current node - which is needed for the check in (2) above. My first approach involved patching a _lot_ of node types and looked a bit like this:

<details>
```javascript
pp.checkIllegalAwaitAndRethrow = function (err, node) {
  if (this.state.potentialIllegalAwaitAt >= node.start) {
    this.raise(this.state.potentialIllegalAwaitAt, "await can only be used in async functions");
  } else {
    throw err;
  }
};

pp.semicolonCheckIllegalAwait = function (node) {
  try {
    this.semicolon();
  } catch (err) {
    this.checkIllegalAwaitAndRethrow(err, node);
  }
};

// Then for *many, many* node types do something like this:
pp.parseReturnStatement = function (node) {
// ... snip ...
  if (this.isLineTerminator()) {
    node.argument = null;
  } else {
    node.argument = this.parseExpression();
    this.semicolonCheckIllegalAwait(node); // instead of this.semicolon();
  }
// ... snip ...
};
```
</details>

But it got out of hand _very_ quickly.

So here's my new recipe, plus some questions.

1. Manage a "current node" stack in  `startNode`/`finishNode` (Is this a viable approach? Does something like this already exist?)
2. In `unexpected()` itself, check for `this.state.potentialIllegalAwaitAt >= currentNode.start` and override the message if `true`.

Honestly, it still seems a bit of a heavy solution that I'm not sure is justifiable here. But at least this way the problem gets treated as a "cross cutting concern", with a lot less code.

I'd appreciate any feedback/ideas! :grinning:

**Comment by kaicataldo** [comment]
Almost ready to make a PR for this!

**Comment by kaicataldo** [comment]
@motiz88 Actually, looked at your PR and the solution I came up with is very different. Don't want to step on your toes here, but let me know what you think of this: https://github.com/kaicataldo/babylon/commit/8358cff052622af9529286653f83bafddee90f4c

Maybe I'm missing something, but it seems to get the job done. If you think it looks good, I can make a PR with this - just wanted to check in with you since you've already done work on this.

**Comment by kaicataldo** [comment]
I'm actually going to see if I can extract the reserved identifier checks out so it's easier to see how all this fits together. Running into some challenges with my own solution - it works great when source type is module, but might actually have to do some kind of flag in state when an `await` keyword is found for when the source type is script, since identifiers are allowed to be called `await`. The easiest thing is to have smarter warnings for source type module and keep the same warnings for source type script.

**Comment by motiz88** [comment]
Hey @kaicataldo, apparently GitHub lost a reply I sent via e-mail so I'll try to address everything again.

For one thing, don't worry about my toes :smile: Especially since my PR is a work-in-progress and in a bit of stasis at the moment. But I think allowing `await` only as the LHS of AssignmentExpression is incorrect, as you've already noticed.

My own solution can probably be simplified and made more correct with the use of `lookahead()`, which I wasn't actually aware of when I started. I might not get back to it for a while though, so feel free to clone that branch and pull anything of interest from it. e.g.

>some kind of flag in state when an await keyword is found 

is exactly what I did.

**Comment by motiz88** [comment]
Hey @kaicataldo, don't worry about my toes :smile: But doesn't your current
solution preclude other valid uses of `await` as an identifier? For example
`await.foo`, `await[foo]`, `await / foo`, `await(foo)`, etc. There's a
partial list in my PR.
